src/core/controllers/oidc.go

```diff
@@ -122,30 +123,100 @@ func (oc *OIDCController) Callback() {
 	}
 	oc.SetSession(tokenKey, tokenBytes)
 
+	oidcSettings, err := config.OIDCSetting()
+	if err != nil {
+		oc.SendInternalServerError(err)
+		return
+	}
+
 	if u == nil {
-		oc.SetSession(userInfoKey, string(ouDataStr))
-		oc.Controller.Redirect(fmt.Sprintf("/oidc-onboard?username=%s", strings.Replace(info.Username, " ", "_", -1)),
-			http.StatusFound)
-	} else {
-		gids, err := group.PopulateGroup(models.UserGroupsFromName(info.Groups, common.OIDCGroupType))
-		if err != nil {
-			log.Warningf("Failed to populate groups, error: %v, user will have empty group list, username: %s", err, info.Username)
-		}
-		u.GroupIDs = gids
-		oidcUser, err := dao.GetOIDCUserByUserID(u.UserID)
-		if err != nil {
-			oc.SendInternalServerError(err)
+		// Recover the username from d.Username by default
+		username := info.Username
+
+		// Fix blanks in username
+		username = strings.Replace(username, " ", "_", -1)
+
+		// If automatic onboard is enabled, skip the onboard page
+		if oidcSettings.AutoOnboard {
+			log.Debug("Doing automatic onboarding\n")
+			user, onboarded := userOnboard(oc, info, username, tokenBytes)
+			if onboarded == false {
```

Comment:
I don't think this is right, in this case the `/callback` will return 200 and nothing happens, we need to explicitly let user know something is wrong.

Maybe when the `info` does not have `username` or user failed to onboard, log the error and redirect to the onboard dialog so user can input the username and onboard?

Reply:
You are right, but I would prefer to use the oc.SendInternalServerError method instead. If you enable AutoOnboard, definitely you don't want to give the user the change to type a username. So in case the automatic onboarding fails I think informing with an HTTP 500 error makes more sense. Let me test this case and see what happens, then provide a better error handling.

Reply:
After some tests, the "userOnboard" function is already handling the error and displaying an error via oc.RenderError (for duplicate user) or just an Internal Server Error for other cases.